### PR TITLE
Support version in config

### PR DESCRIPTION
**Pull request by live1206**
[AB#195159](https://dev.azure.com/ceapex/Engineering/_workitems/edit/195159)

- support `version` in `docfx.json` of v2
- `version` takes precedence over `group`, if both configured
- add tests for versioning scenarios with v2 config

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/5741)

### CODE REVIEW

File: src/docfx/Errors.cs [hunk 1]
```diff
@@ -430,6 +430,9 @@ public static Error UidPropertyConflict(string uid, string propertyName, IEnumer
 
         public static class Versioning
         {
+            public static Error DuplicateVersionConfig(SourceInfo? source)
```
Comment (yufeih) [inline src/docfx/Errors.cs:433]: consider remove this error code.

File: src/docfx/build/moniker/MonikerProvider.cs [hunk 2]
```diff
@@ -17,8 +17,8 @@ internal class MonikerProvider
 
         private readonly (Func<string, bool> glob, SourceInfo<string?>)[] _rules;
 
-        private readonly ConcurrentDictionary<FilePath, SourceInfo<string?>> _monikerRangeCache
-                   = new ConcurrentDictionary<FilePath, SourceInfo<string?>>();
+        private readonly ConcurrentDictionary<FilePath, (List<Error>, SourceInfo<string?>, string[])> _monikerRangeCache
```
Comment (yufeih) [inline src/docfx/build/moniker/MonikerProvider.cs:20]:
```suggestion
        private readonly ConcurrentDictionary<FilePath, (List<Error>, SourceInfo<string?>)> _monikerRangeCache
```

File: src/docfx/build/moniker/MonikerProvider.cs [hunk 3]
```diff
@@ -40,14 +40,8 @@ public MonikerProvider(Config config, BuildScope buildScope, MetadataProvider me
             _rules = _config.MonikerRange.Select(pair => (GlobUtility.CreateGlobMatcher(pair.Key), pair.Value)).Reverse().ToArray();
         }
 
-        public SourceInfo<string?> GetConfigMonikerRange(FilePath file)
+        public (List<Error>, SourceInfo<string?>, string[]) GetConfigMonikerRange(FilePath file)
```
Comment (yufeih) [inline src/docfx/build/moniker/MonikerProvider.cs:43]:
```csharp
if (_hasConfigMonikerRange)
{
}
```

File: src/docfx/build/moniker/MonikerProvider.cs [hunk 4]
```diff
@@ -156,24 +150,43 @@ private static (Error?, string[]) GetMonikerIntersection(UserMetadata metadata,
             return (error, intersection);
         }
 
-        private SourceInfo<string?> GetConfigMonikerRangeCore(FilePath file)
+        private (List<Error>, SourceInfo<string?>, string[]) GetConfigMonikerRangeCore(FilePath file)
```
Comment (yufeih) [inline src/docfx/build/moniker/MonikerProvider.cs:153]:
```suggestion
        private (SourceInfo<string?>) GetConfigMonikerRangeCore(FilePath file)
```

File: docs/specs/moniker.yml [hunk 5]
```diff
@@ -1808,3 +1808,184 @@ inputs:
 outputs:
   136a42ac/docs/a.json: |
     {"metadata":{"monikers":["netcore-1.0"]}}
+---
+# Compatible for group in docfx.json of v2
+repos:
+  https://github.com/ops/config-1:
+  - files:
+      .openpublishing.publish.config.json: |
```
Comment (yufeih) [inline docs/specs/moniker.yml:1816]: consider creating a minimal test spec, `.openpublishing.publish.config.json` is not needed? Same as other test cases.

File: docs/specs/moniker.yml [hunk 6]
```diff
@@ -1808,3 +1808,184 @@ inputs:
...
+        {}
+      docfx.json: |
```
Comment (yufeih) [inline docs/specs/moniker.yml:1818]: consider use YAML to make config easier to read
```suggestion
      docfx.yml: |
```

File: docs/specs/moniker.yml [hunk 7]
```diff
@@ -1808,3 +1808,184 @@ inputs:
 outputs:
   136a42ac/docs/a.json: |
     {"metadata":{"monikers":["netcore-1.0"]}}
+---
+# Compatible for group in docfx.json of v2
+repos:
```
Comment (yufeih) [inline docs/specs/moniker.yml:1813]:
```suggestion
inputs:
```

File: docs/specs/moniker.yml (hunk 5 above)
Comment (live1206) [reply docs/specs/moniker.yml:1816]: sure, thought we need to have `.openpublishing.publish.config.json` along with v2 `docfx.json`

File: docs/specs/moniker.yml (hunk 6 above)
Comment (live1206) [reply docs/specs/moniker.yml:1818]: Time-consuming to convert it to `YAML`, will stick to `JSON` here...

**Pull request merged**